Updating catbox to hapi 17 bindings.

saboya opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Frees hapi 17 from Boomv4 dependencies, along with all the packages that depend on it.

SimonSchick commented on types/hapi-auth-basic/tsconfig.json:
```diff
@@ -17,6 +17,9 @@
                 "boom/v4"
             ],
+            "catbox": [
+                "catbox/v7"
```
Need to look into updating those packages in the future as well 😄

saboya replied:
Well, there's the PR pending for hapi-auth-basic and nes :) I'll rebase it after this one.

SimonSchick commented on types/catbox/index.d.ts:
```diff
@@ -21,36 +18,39 @@ export type CallBackWithResult<T> = (err: Boom.BoomError | null | undefined, res
-    isReady(): boolean;
-    /** validateSegmentName(segment) - returns null if the segment name is valid (see below), otherwise should return an instance of Error with an appropriate message. */
-    validateSegmentName(segment: string): null | Boom.BoomError;
+  constructor(engine: EnginePrototypeOrObject, options: ClientOptions);
```
Why change the indenting? It makes the diff hard to read...

saboya replied:
Probably because I copy pasted this file and the editor auto-formatted it. Will fix and squash.

SimonSchick commented on types/catbox/index.d.ts:
```diff
@@ -21,36 +18,39 @@ export type CallBackWithResult<T> = (err: Boom.BoomError | null | undefined, res
+  isReady(): boolean;
+
+  /** validateSegmentName(segment) - returns null if the segment name is valid (see below), otherwise should return an instance of Error with an appropriate message. */
+  validateSegmentName(segment: string): null | Error;
```
Why change it to `Error` instead of `BoomError`?

saboya replied:
The API makes no mention of `Boom` errors, and even if the code suggests they will all be `Boom` instances, I always tend to go with the API reference:

https://github.com/hapijs/catbox

Also, as of Boom v5, there's no `BoomError` anymore, the equivalent would be `Boom.Boom`.

saboya replied:
(and Boom.Boom extends Error)

SimonSchick replied:
Yea that's good enough for me.

saboya commented:
So, I just rebased to latest master, to keep up with https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23735 and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23737, however, I believe there's a test in hapi/v16 that should be removed.

`types/hapi/v16/response/error-representation.ts` has a dependency on `vision`, which makes all the packages that depend on `hapi` pull in that too. If that's the case, basically everywhere I just added `catbox/v7` needs `vision/v4` added as well, and since `hapi` doesn't actually depend on `vision`, I just doesn't seem right.

Thoughts?

SimonSchick commented:
I think it's fair to remove that test then, I doubt the old typings will be updated.

saboya commented:
Done. It isn't really an issue with the `catbox` update, but figured it would be faster than doing another PR. Tests should be passing now.

saboya commented:
Done.
